src/jdk.compiler/share/classes/com/sun/tools/javac/util/StringNameTable.java

```diff
@@ -0,0 +1,139 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.tools.javac.util;
+
+import java.util.HashMap;
+
+/**
+ * Straightforward implementation of {@link Name.Table} using {@link String}s.
+ *
+ *  <p><b>This is NOT part of any supported API.
+ *  If you write code that depends on this, you do so at your own risk.
+ *  This code and its internal interfaces are subject to change or
+ *  deletion without notice.</b>
+ */
+public class StringNameTable extends Name.Table {
+
+    private final HashMap<String, Name> nameMap;
+
+// Factory
+
+    public static StringNameTable create(Names names) {
+        return new StringNameTable(names);
+    }
+
+// Constructors
+
+    public StringNameTable(Names names) {
+        this(names, 8000);
+    }
+
+    public StringNameTable(Names names, int initialCapacity) {
+        super(names);
+        this.nameMap = new HashMap<>(initialCapacity);
+    }
+
+// Name.Table
+
+    @Override
+    public Name fromString(String string) {
+        return this.nameMap.computeIfAbsent(string, s -> new NameImpl(this, s));
+    }
+
+    @Override
+    public Name fromChars(char[] buf, int off, int len) {
+        return this.fromString(new String(buf, off, len));
+    }
+
+    @Override
+    public Name fromUtf(byte[] buf, int off, int len, Convert.Validation validation) throws InvalidUtfException {
+        return this.fromString(Convert.utf2string(buf, off, len, validation));
+    }
+
+    @Override
+    public void dispose() {
+        this.nameMap.clear();
+    }
+
+// NameImpl
+
+    private static final class NameImpl extends Name {
+
+        private final String string;
+
+    // Constructor
+
+        NameImpl(StringNameTable table, String string) {
+            super(table);
+            this.string = string;
+        }
+
+        @Override
+        public String toString() {
+            return string;
+        }
+
+        @Override
+        public boolean contentEquals(CharSequence cs) {
+            return string.contentEquals(cs);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this)
+                return true;
+            if (obj == null || obj.getClass() != getClass())
+                return false;
+            final NameImpl that = (NameImpl)obj;
+            return that.table == table && that.string.equals(string);
```

Comment:
This can use `instanceof <pattern>` to match `Utf8NameTable.NameImpl::equals`:
```suggestion
            return (obj instanceof NameImpl name)
                && table == name.table
                && string.equals(name.string);
```

Reply:
Thanks. I agree the two implementations should be consistent. However you also made me realize that some common logic (e.g., type and same table checks) can be pushed up into the superclass. Also I think we should keep the initial same object check though for performance.

Refactored in c5a2546e8a5.